README.md
# WISTER

```
 __          _______  _____ _______ ______ _____  
 \ \        / /_   _|/ ____|__   __|  ____|  __ \ 
  \ \  /\  / /  | | | (___    | |  | |__  | |__) |
   \ \/  \/ /   | |  \___ \   | |  |  __| |  _  / 
    \  /\  /   _| |_ ____) |  | |  | |____| | \ \ 
     \/  \/   |_____|_____/   |_|  |______|_|  \_\

                                     Cycurity       
```

## SUMMARY

> A unique wordlist generator with several types of combinations to choose from, or to mix them all. Capable of receiving words from various sources and outputing them to different encoding types.

## USAGE

![](https://raw.githubusercontent.com/cycurity/wister/main/running.gif "Usage")

```
usage: wister.py [-h] [-c [1-5] [[1-5] ...]] [-d [2-5]] [-e FORMAT] [-i FILE] [-l] [-m [0-49]] [-M [1,50]] [-o FILE] [-N] [-v] [-V] [-w WORDS [WORDS ...]]

options:
  -h, --help            show this help message and exit
  -c [1-5] [[1-5] ...], --combination [1-5] [[1-5] ...]
                        Select which types of combinations to use.
  -d [2-5], --depth [2-5]
                        Select the depth of word mixing when using combination mode 1. Default: 3
  -e FORMAT, --encode FORMAT
                        Encode/Hash the generated wordlist with a given algorithm: "md5", "base64", "hex", "url", "sha1", "sha256" , "sha512", "sha3-256", "sha3-512"
  -i FILE, --input FILE
                        Select the input file containing the words.
  -l, --list            List the types of combination.
  -m [0-49], --min [0-49]
                        Select minimum characters for output. Default: 0
  -M [1,50], --max [1,50]
                        Select maximum characters for output. Default: 20
  -o FILE, --output FILE
                        Specify the output file. Default: output.lst
  -N, --noprogress      Disables progress bar (increases performance)
  -v, --version         Returns the version of the program.
  -V, --verbose         Verbose the output.
  -w WORDS [WORDS ...], --words WORDS [WORDS ...]
                        Words to generate the output with.
```

## INSTALLATION

The process to install Wister is very simple, just copy the following commands:

PIP module installation:

```bash
$ sudo apt-get install python3 python3-pip
$ pip install wister
```

Source compilation:

```bash
$ sudo apt-get install python3 python3-pip git
$ git clone https://github.com/cycurity/wister.git
$ cd wister
$ pip install -r requirements.txt
$ pip install .
```

## MANUAL

&nbsp;&nbsp;&nbsp;&nbsp;To run the program, you need to pass at least a set of words, using an input file (`-i` or `--input`), or passing the words manually using the `-w` or `--words` tag. If no more options are passed, the program will just write the specified words to the default output file `output.lst`.

&nbsp;&nbsp;&nbsp;&nbsp;There are 5 types of combinations that can be used (with `-c` or `--combination`) simultaneously : `1. Word Mix`; `2. Case Alternate`; `3. Homograph`; `4. Reverser`; `5. Saltify`. To get more information about each type of combination, you should use the *list* argument (`-l` or `--list`). When using the first combination (Word Mix), you can also select the depth (`-d` / `--depth`) which will define the number of words being mixed together. By default, this value is set to 3. 

&nbsp;&nbsp;&nbsp;&nbsp;Addicionally, you can define the maximum generated word length (`-M` or `--max`) or the minimum number of characters required (`-m` or `--min`). By default, the minimum value is set to 0 and the max to 20.

&nbsp;&nbsp;&nbsp;&nbsp;To specify the desired output file, you must use `-o` or `--output`, followed by the filename.

&nbsp;&nbsp;&nbsp;&nbsp;The verbose option (`-V` / `--verbose`) will allow you to follow the generation of the wordlist, by printing the several steps of the program. When this option is enabled, the progress bar is disabled.

&nbsp;&nbsp;&nbsp;&nbsp;Optionally, the generated wordlist can be encoded or hashed, using `-e` or `--encode`, with specific types of algorithms: `md5`, `base64`, `hex`, `url`, `sha1`, `sha256`, `sha512`, `sha3-256`, `sha3-512`.

&nbsp;&nbsp;&nbsp;&nbsp;By default, the progress bar is enabled when verbose mode is not. You can also choose to disable it using the `-N` or `--noprogress` flag. Disabling this feature will increase performance, decreasing the overall execution time of the program.

&nbsp;&nbsp;&nbsp;&nbsp;To print the current program version, use `-v` or `--version`.

&nbsp;&nbsp;&nbsp;&nbsp;To get the usage and help page, use the `-h` flag or `--help`.

## CREDITS

Developed by:

- [fssecur3](https://github.com/fssecur3 "fssecur3's Github Profile")
- [intMa1n](https://github.com/Bernardo15Sousa "intMa1n's Github Profile")

Project maintained by:

- [fssecur3](https://github.com/fssecur3 "fssecur3's Github Profile")

## LICENSE

Copyright (C) 2022, Cycurity

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

The software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

Check out the GNU General Public License: [http://www.gnu.org/licenses/](http://www.gnu.org/licenses/)


CHANGELOG.md
# WISTER - Changelog

## *1.0* - 01/04/2022 - First release

## *1.0.1* - 05/04/2022 - Generated output file size

- Alert the file size from the generated output file to the user when the save ends

## *1.0.2* - 18/04/2022 - Added small feature to mode 1 - Word Mix

- Original words will also repeat themselves (i.e.: hello -> hellohello) once.

## *1.0.3* - 10/08/2022 - Project restructure and small fix

- Restructured the project
- Created the python module and added it to pip
- Fixed gramatical errors on help and README.md
- Added installation instructions for the new module in README.md
- Added encoding instructions in README.md


wister/utils.py
import argparse
import sys
import hashlib
import urllib.parse
import base64

def args(version):
    parser = argparse.ArgumentParser(add_help=True)
    parser.add_argument("-c", "--combination", dest="types", metavar="[1-5]", choices=(1,2,3,4,5), nargs="+", type=int, help="Select which types of combinations to use.")
    parser.add_argument("-d", "--depth", metavar="[2-5]", choices=(2,3,4,5), type=int, default=3, help="Select the depth of word mixing when using combination mode 1. Default: 3")
    parser.add_argument("-e", "--encode", metavar="FORMAT", choices=("md5", "base64", "hex", "url", "sha1", "sha256" , "sha512", "sha3-256", "sha3-512"), help='Encode/Hash the generated wordlist with a given algorithm: "md5", "base64", "hex", "url", "sha1", "sha256" , "sha512", "sha3-256", "sha3-512"')
    parser.add_argument("-i", "--input", metavar="FILE", type=argparse.FileType('r'), dest="input", help="Select the input file containing the words.")
    parser.add_argument("-l", "--list", action="store_true", help="List the types of combination.")
    parser.add_argument("-m", "--min", metavar="[0-49]", default=0, type=int, help="Select minimum characters for output. Default: 0")
    parser.add_argument("-M", "--max", metavar="[1,50]", default=20, type=int, help="Select maximum characters for output. Default: 20")
    parser.add_argument("-o", "--output", default="output.lst", metavar="FILE", type=argparse.FileType('w'), dest="output", help="Specify the output file. Default: output.lst")
    parser.add_argument("-N", "--noprogress", dest="noprog", action="store_true", default=False, help="Disables progress bar (increases performance)")
    parser.add_argument("-v", "--version", action="version", version=f"{version}", help="Returns the version of the program.")
    parser.add_argument("-V", "--verbose", action="store_true", default=False, help="Verbose the output.")
    parser.add_argument("-w", "--words", nargs='+', type=str, help="Words to generate the output with.")
    
    # ADD argument for number of threads to execute
    # ADD argument to ask if the user wants the generated words to be echoed on stdout
    return parser.parse_args()

def list():
    print("""
    Types of Combinations

1.  Word Mix - The words are mixed between each other
        abc123
        admin123
        123abc
        123admin
        abcadmin
        admin123abc
        ...

2.  Case Alternate - The case is mixed for each word or combination
        Admin
        admiN
        AdmiN
        ADMIN

3.  Homograph - Change specific characters for others, commonly used, similar ones
        4dmin
        @dmin
        adm1n
        adm!n
        4dm1n
        @dm1n
        ...

4.  Reverser - Reverse the words
        admin
        123
        abc
        nimda
        321
        cba

5.  Saltify - Add some salt (like ponctuation or numbers)
        admin1
        admin.123
        abc_admin
        123-abc
        ...

You can add multiple types of combinations when executing.""")

def progress(progress):
    barLength = 40
    if isinstance(progress, int):
        progress = float(progress)
    if progress < 0:
        progress = 0
    if progress >= 1:
        progress = 1
    block = int(round(barLength * progress))
    text = "\r[{0}] {1:.0f}%".format( "#" * block + "-" * (barLength - block), progress * 100)
    sys.stdout.write(text)
    sys.stdout.flush()

def wordfile (file):
    words = []
    for word in file:
        word = word.strip()
        words.append(word)
    file.close()
    return words

def output (file, words: list, v):
    if (v):
        print("Saving to '" + file.name + "'...")
    for i in words:
        file.write(i + "\n")
    bts = file.tell()
    file.close()
    return bts

def encode(final, encoding):
    newfinal = []
    if (encoding == "md5"):
        for w in final:
            newfinal.append(hashlib.md5(w.encode()).hexdigest())
    elif (encoding == "url"):
        for w in final:
            newfinal.append(urllib.parse.quote_plus(w))
    elif (encoding == "base64"):
        for w in final:
            newfinal.append(str(base64.b64encode(w.encode("utf-8")), "utf-8"))
    elif (encoding == "hex"):
        for w in final:
            newfinal.append(w.encode("utf-8").hex())
    elif (encoding == "sha1"):
        for w in final:
            newfinal.append(hashlib.sha1(w.encode()).hexdigest())
    elif (encoding == "sha256"):
        for w in final:
            newfinal.append(hashlib.sha256(w.encode()).hexdigest())
    elif (encoding == "sha512"):
        for w in final:
            newfinal.append(hashlib.sha512(w.encode()).hexdigest())
    elif (encoding == "sha3-512"):
        for w in final:
            newfinal.append(hashlib.sha3_512(w.encode()).hexdigest())
    else:
        for w in final:
            newfinal.append(hashlib.sha3_256(w.encode()).hexdigest())
    return newfinal

def convert(b):
    b = int(b)
    if (b > 1000000000):
        return f"{b / 1000000000:.2f} Gb"
    elif (b > 1000000):
        return f"{b / 1000000:.2f} Mb"
    elif (b > 1000):
        return f"{b / 1000:.2f} Kb"
    else:
        return f"{b:.2f} B"